pkg/pillar/cmd/baseosmgr/handlebaseos.go

```diff
@@ -1018,3 +1021,107 @@ func handleOtherPartRebootReason(ctxPtr *baseOsMgrContext, status *types.BaseOsS
 		status.SetError(reason, ctxPtr.rebootTime)
 	}
 }
+
+//isImageInErrorState returns true if we try to update to IMGB without success
+//also returns ZbootStatus
+func isImageInErrorState(ctxPtr *baseOsMgrContext) (bool, *types.ZbootStatus) {
+	otherPartName := zboot.GetOtherPartition()
+	partStatus := getZbootStatus(ctxPtr, otherPartName)
+	if partStatus == nil {
+		log.Functionf("No other partition status for %s",
+			otherPartName)
+		return false, nil
+	}
+	shortVerOtherPart := partStatus.ShortVersion
+	if shortVerOtherPart == "" {
+		log.Functionf("Other partition has no version")
+		return false, partStatus
+	}
+	if partStatus.PartitionState != "inprogress" {
+		log.Functionf("Other partition state %s not inprogress",
+			partStatus.PartitionState)
+		return false, partStatus
+	}
+	baseOSConfig := lookupBaseOsConfigByVersion(ctxPtr, shortVerOtherPart)
+	if baseOSConfig == nil {
+		log.Functionf("Cannot found BaseOsConfig with %s version; ignoring RetryUpdateCounter",
+			shortVerOtherPart)
+		return false, partStatus
+	}
+	if !baseOSConfig.Activate {
+		log.Functionf("BaseOsConfig %s has no activate; ignoring RetryUpdateCounter", baseOSConfig.Key())
+		return false, partStatus
+	}
+	return true, partStatus
+}
+
+// handleUpdateRetryCounter checks
+// if current partition is not active: just update configUpdateRetry
+// if other partition is in failed state: if retryUpdateCounter changed: save current counters and initiate re-update
+// in other case update configUpdateRetry and currentUpdateRetry, save them and publish
+func handleUpdateRetryCounter(ctxPtr *baseOsMgrContext, retryUpdateCounter uint32) {
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Warnf("handleUpdateRetryCounter: No current partition status for %s; failed RetryUpdateCounter",
+			curPartName)
+		return
+	}
+	if partStatus.PartitionState != "active" {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
```

Comment:
Shouldn't you save the updated counter here?

Reply:
It will slightly change the logic. In this case when we are in testing, and receive new counter, we will save it, and than if we are failing, we will reboot and pick up saved counter. But with the saved counter we will not affected with check below and will not do re-update.

Reply:
Maybe comments and/or logs should say that we are in testing. But also shouldn't we completely ignore it? Seems odd to have different counter in memory and on disk; makes the case analysis a lot more complex.

Reply:
I put log here and remove updating of value in-memory